module-system/module-implementations/sov-blob-storage/src/capabilities.rs

```diff
@@ -16,36 +20,88 @@ impl<C: Context, Da: DaSpec> BlobSelector<Da> for BlobStorage<C, Da> {
     where
         I: IntoIterator<Item = &'a mut Da::BlobTransaction>,
     {
+        // Calculate any expiring deferred blobs first, since these have to be processed no matter what (Case 2 above).
+        // Note that we have to handle this case even if there is no preferred sequencer, since that sequencer might have
+        // exited while there were defferred blobs waiting to be processed
         let current_slot: TransitionHeight = self.get_current_slot_height(working_set);
-        let past_deferred: Vec<Da::BlobTransaction> = current_slot
-            .checked_sub(self.get_deferred_slots_count(working_set))
-            .map(|pull_from_slot| self.take_blobs_for_slot_height(pull_from_slot, working_set))
-            .unwrap_or_default();
-        let preferred_sequencer = self.get_preferred_sequencer(working_set);
+        let slot_for_expiring_blobs =
+            current_slot.saturating_sub(self.get_deferred_slots_count(working_set));
+        let expiring_deferred_blobs: Vec<Da::BlobTransaction> =
+            self.take_blobs_for_slot_height(slot_for_expiring_blobs, working_set);
 
-        let preferred_sequencer = if let Some(sequencer) = preferred_sequencer {
+        // If there is no preferred sequencer, that's all we need to do
+        let preferred_sequencer = if let Some(sequencer) = self.get_preferred_sequencer(working_set)
+        {
             sequencer
         } else {
             // TODO: https://github.com/Sovereign-Labs/sovereign-sdk/issues/654
             // Prevent double number of blobs being executed
-            return Ok(past_deferred
+            return Ok(expiring_deferred_blobs
                 .into_iter()
                 .map(Into::into)
                 .chain(current_blobs.into_iter().map(Into::into))
                 .collect());
         };
 
+        // If we reach this point, there is a preferred sequencer, so we need to handle cases 1 and 3.
+
+        // First, compute any "bonus blobs" requested
+        // to be processed early.
+        let num_bonus_blobs_requested = self
+            .deferred_blobs_requested_for_execution_next_slot
+            .get(working_set)
+            .unwrap_or_default();
+        self.deferred_blobs_requested_for_execution_next_slot
+            .set(&0, working_set);
+
+        let mut remaining_blobs_requested =
+            (num_bonus_blobs_requested as usize).saturating_sub(expiring_deferred_blobs.len());
+        let mut bonus_blobs: Vec<BlobRefOrOwned<Da::BlobTransaction>> =
+            Vec::with_capacity(remaining_blobs_requested);
+        let mut next_slot_to_check = slot_for_expiring_blobs + 1;
+        while remaining_blobs_requested > 0 && next_slot_to_check < current_slot {
```

Comment:
I was initially thought why not to get up to current blob including, but then later realizaed why.
Maybe let's put this comment. Feel free to edit the wording.

```suggestion
        // Blobs from the current slot are excluded, ensuring that priority blobs are not deprioritized.
        while remaining_blobs_requested > 0 && next_slot_to_check < current_slot {
```

Reply:
Blobs from the current slot are excluded because they haven't been stored yet, not because that would cause priority blobs to be deprioritized. Adding a comment to clarify.